### ISSUE TITLE
disable_adapter() / enable_adapter_layers() silently unfreezes adapters loaded with inference_mode=True

### ISSUE THREAD

**Issue by ErenAta16** [huggingface/peft#3447]
### System Info

Reproduced against the current `main` branch source (`src/peft/tuners/tuners_utils.py`, `BaseTunerLayer.enable_adapters`), with `torch` installed locally, `get_peft_model` + `LoraConfig`.

### Who can help?

@benjaminbossan @githubnemo

### Reproduction

```python
import torch
import torch.nn as nn
from peft import get_peft_model, LoraConfig


class SmallModel(nn.Module):
    def __init__(self):
        super().__init__()
        self.lin = nn.Linear(8, 8)

    def forward(self, x):
        return self.lin(x)


model = get_peft_model(
    SmallModel(),
    LoraConfig(target_modules=["lin"], r=4, lora_alpha=8, inference_mode=True),
)

def lora_requires_grad():
    return [p.requires_grad for n, p in model.named_parameters() if "lora_" in n]

print("Before:", lora_requires_grad())

with model.disable_adapter():
    print("Inside:", lora_requires_grad())

print("After:", lora_requires_grad())
```

Actual output:
```
Before: [False, False]
Inside: [False, False]
After: [True, True]
```

### Expected behavior

`requires_grad` should stay `False` after the `with model.disable_adapter():` block exits, matching the `inference_mode=True` the adapter was loaded with. Instead it flips to `True`, silently unfreezing an adapter that was explicitly configured as frozen.

### Root cause

`BaseTunerLayer.enable_adapters` (`src/peft/tuners/tuners_utils.py`, lines 1564-1574):

```python
def enable_adapters(self, enabled: bool) -> None:
    """Toggle the enabling and disabling of adapters
    Takes care of setting the requires_grad flag for the adapter weights.
    """
    if enabled:
        self.set_adapter(self.active_adapters)
        self._disable_adapters = False
    else:
        ...
```

`set_adapter` is called with no `inference_mode` argument, so it defaults to `False`, forcing `requires_grad=True` on every active adapter parameter regardless of the config the adapter was actually loaded/set with.

`PeftModel.disable_adapter()` (`src/peft/peft_model.py`, ~lines 1068-1084) calls `self.base_model.enable_adapter_layers()` in its `finally` block to restore adapters after temporarily disabling them, the standard pattern for computing a reference/base-model output (e.g. in DPO/PPO/RLHF-style training scripts). Every use of this context manager unconditionally re-enables `requires_grad=True` on the active adapter, even when it was intentionally frozen.

### Not a duplicate of #2928 / PR #2936

That issue/PR is about `modules_to_save`/`AuxiliaryTrainingWrapper` specifically (the *original*, non-adapter module's gradient state getting toggled), and its own fix only touches `_set_adapter` in `utils/other.py`. It doesn't touch `BaseTunerLayer.enable_adapters` in `tuners_utils.py`, the path exercised here, and the existing test added for that PR (`test_requires_grad_modules_to_save_disabling`) never constructs an adapter with `inference_mode=True`, so this path isn't covered.

### Impact

Any code that calls `model.disable_adapter()` (or `enable_adapter_layers()` directly) on a model containing an adapter that was explicitly frozen (`inference_mode=True`, e.g. a loaded reference adapter for RLHF-style training, or any adapter loaded with `is_trainable=False`) will have that adapter silently become trainable again as soon as the context manager exits. This can corrupt a "frozen reference" adapter if an optimizer or trainable-parameter filter is built after such a call, and makes `requires_grad` an unreliable signal for introspecting which parameters are actually meant to be trainable.

**Comment by DaoyuanLi2816** [comment]
Confirmed — ran your repro against a fresh `origin/main` checkout and got the identical output (`Before: [False, False]`, `Inside: [False, False]`, `After: [True, True]`). The root cause diagnosis is correct: `BaseTunerLayer.enable_adapters` calls `self.set_adapter(self.active_adapters)` with no `inference_mode` argument, so it defaults to `False` and unconditionally sets `requires_grad=True`, regardless of the config the active adapter(s) were actually loaded/set with. Also confirmed this isn't a duplicate of #2928/PR #2936 or of PR #3290 — those touch different call sites (`_set_adapter`/`ModulesToSaveWrapper.set_adapter`, and `BaseTuner.inject_adapter`'s housekeeping call, respectively), not `BaseTunerLayer.enable_adapters`.

While tracing this, I found the identical unconditional-unfreeze pattern independently duplicated in two more places, so `modules_to_save` and LN Tuning adapters have the same bug (`TrainableTokensWrapper`/`UniLoraLayer` are also affected, but only because they delegate to a `BaseTunerLayer`-derived object, so they're fixed by the same root-cause change):

```python
config = LoraConfig(target_modules=["lin"], modules_to_save=["cls"], r=4, inference_mode=True)
model = get_peft_model(SmallModel(), config)
print([p.requires_grad for n, p in model.named_parameters() if "modules_to_save" in n])  # [False, False]
with model.disable_adapter():
    pass
print([p.requires_grad for n, p in model.named_parameters() if "modules_to_save" in n])  # [True, True] -- same bug
```

I intend to open a PR shortly that fixes `BaseTunerLayer.enable_adapters` (and the two duplicated copies) by having `BaseTuner._enable_adapter_layers` look up each active adapter's own `inference_mode` from `peft_config` when re-enabling, instead of unconditionally marking it trainable — mirroring how `inject_adapter`'s housekeeping call already does this lookup. Will add regression tests parametrized across the existing `TEST_CASES` matrix plus dedicated `modules_to_save`/`trainable_token_indices` tests, following TDD (failing test first) per the contribution guide. Posting this comment first per the repo's AI-assisted contribution policy before putting up the PR.

**Comment by BenjaminBossan** [comment]
We are aware of this behavior and intend to fix it, but it's backwards incompatible and as such we can't change it right now. We might address the behavior in a possible PEFT v1.0 release.

**Comment by ErenAta16** [comment]
Understood, thanks for the context on the backwards-compatibility constraint. Makes sense to hold this for a v1.0-style break rather than changing default behavior under people's feet in a patch release.

Appreciate DaoyuanLi2816 double-checking the root cause and mapping out the other two call sites with the same pattern (`modules_to_save`, LN Tuning) too, that'll save whoever picks this up in v1.0 some time.
